File: built-in plugins dsp/mb-src.cpp
/**
 * MB Sample Rate Converter
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : High-quality sample rate conversion with anti-aliasing
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SRC_H
#define MB_SRC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSrc : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-src";
    static constexpr const char* PLUGIN_NAME    = "MB Sample Rate Converter";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float targetRate = 44100f;  // range [22050, 192000]
    float quality = 2f;  // range [0, 3]
    float antiAlias = 1f;  // range [0, 1]
    };

    MbSrc() = default;
    ~MbSrc() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.targetRate = std::clamp(params.targetRate, 22050f, 192000f);
        params.quality = std::clamp(params.quality, 0f, 3f);
        params.antiAlias = std::clamp(params.antiAlias, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Sample Rate Converter
        return input;
    }
};

#endif // MB_SRC_H

File: built-in plugins dsp/mb-src.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSrcPlugin: PluginDefinition = { id: 'mb-src', slug: 'mb-src', name: 'MB Sample Rate Converter', category: 'effect', type: 'mastering' as any, version: '1.0.0', description: 'High-quality sample rate conversion with anti-aliasing', author: 'Max Booster', parameters: [{ id: 'targetRate', name: 'Target Rate', type: 'float', defaultValue: 44100, minValue: 22050, maxValue: 192000, automatable: false }, { id: 'quality', name: 'Quality', type: 'float', defaultValue: 2, minValue: 0, maxValue: 3, automatable: false }, { id: 'antiAlias', name: 'Anti-Alias', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: false }], defaultPreset: { targetRate: 44100, quality: 2, antiAlias: 1 } };

export default MbSrcPlugin;
